sdk/identity/azure-identity/src/default_azure_credential.cpp

```diff
@@ -0,0 +1,72 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include "azure/identity/default_azure_credential.hpp"
+
+#include "azure/identity/azure_cli_credential.hpp"
+#include "azure/identity/environment_credential.hpp"
+#include "azure/identity/managed_identity_credential.hpp"
+
+#include "azure/core/internal/diagnostics/log.hpp"
+
+using namespace Azure::Identity;
+using namespace Azure::Core::Credentials;
+
+using Azure::Core::Context;
+using Azure::Core::Diagnostics::Logger;
+using Azure::Core::Diagnostics::_internal::Log;
+
+namespace {
+std::string const IdentityPrefix = "Identity: ";
+}
+
+DefaultAzureCredential::DefaultAzureCredential(TokenCredentialOptions const& options)
+{
+  // Initializing m_credential below and not in the member initializer list to have a specific order
+  // of log messages.
+  auto const logLevel = Logger::Level::Verbose;
+  if (Log::ShouldWrite(logLevel))
+  {
+    Log::Write(
+        logLevel,
+        IdentityPrefix
+            + "Creating DefaultAzureCredential which combines mutiple parameterless credentials "
+              "into a single one (by using ChainedTokenCredential)."
+              "\nDefaultAzureCredential is only recommended for the early stages of development, "
+              "and not for usage in production environment."
+              "\nOnce the developer focuses on the Credentials and Authentication aspects of their "
+              "application, DefaultAzureCredential needs to be replaced with the credential that "
+              "is the better fit for the application.");
+  }
+
+  // Creating credentials in order to ensure the order of log messages.
+  auto const envCred = std::make_shared<EnvironmentCredential>(options);
```

Comment:
That's because ChainedTokenCredential takes `std::list<shared_ptr>` - and if I create them as unique_ptr's, in initialization just a few lines below I'll need to call `release()` anyway. Const shared_ptrs can't be `reset()` to point to another object, but they can be shared, it calls copy constructor which increases the counter.